### PR TITLE
fix(graph): validate attributes on the manual entity/relation APIs (GHSA-c922-pw4m-4wcv)

### PR DESCRIPTION
**Pull request by danielaskdd**
## Summary

Closes the unvalidated graph-attribute write reported as GHSA-c922-pw4m-4wcv, and the four further instances of the same defect that the report does not cover.

`aedit_entity` / `aedit_relation` merged the caller's mapping into the stored object wholesale (`{**node_data, **updated_data}`) with no key allowlist and no check on the value, so any caller that can reach `POST /graph/entity/edit` could write an attribute no graph backend is able to store.

## Motivation

On the default NetworkX backend this is not one failed request. `upsert_node` mutates the in-memory `nx.Graph` before serialization and nothing rolls it back, so the unstorable value stays on the node for the life of the process; `write_nx_graph` serializes the *whole* graph, so every later flush by any caller re-hits the same failure. Entity writes and document ingestion fail from then on while reads and `/health` keep returning 200.

Verified against a real `NetworkXStorage` at the storage level (no server, no LLM). Four things came out of that which change the shape of the fix:

1. **Non-scalars are not the only unstorable values.** `None` is refused by GraphML exactly like `dict`, and a string holding a code point XML cannot encode fails identically. An allowlist that admits "any scalar, plus `None`" leaves the original outage fully reachable through an accepted field name — `{"file_path": null}` and a `file_path` carrying `U+000B` both reproduce it.
2. **A storable scalar of the wrong type is worse, not better.** `{"source_id": 123}` serializes fine (GraphML long, JSON number), reaches disk and **survives a restart**, after which every reader that splits `source_id` on `GRAPH_FIELD_SEP` raises `AttributeError` — including the ingestion merge in `operate.py`. So fields are validated against their own type, not against "is a scalar".
3. **The create path is affected too.** Copying six named fields stops an unknown *key*, but a non-scalar value on a known one reached `upsert_node` unexamined: `POST /graph/entity/create` with `{"entity_data": {"entity_type": {"any": ["object"]}}}` produces the identical instance-wide outage. Only `description` was incidentally safe, because building the VDB content concatenates it.
4. **There is a third copy of the merge.** `_merge_entities_impl` applies `target_entity_data` the same way. Not reachable from the HTTP route (the merge endpoint does not pass it), but part of the public Python API.

The other backends are not exempt, they just fail differently — and mostly worse, because they *persist*: PGTableGraph (`jsonb`) and MongoGraph store a nested object verbatim and return 200; OpenSearchGraph stores it and adds a field to the index mapping; Neo4j/Memgraph refuse it per-request; on the Cypher backends `null` silently *deletes* the property instead of erroring. One request, four outcomes — which is why the rule belongs at the backend-independent entry layer.

## What's changed

**Entry layer** (`lightrag/utils_graph.py`) — `_sanitize_graph_fields` at the five public entry points (`aedit_entity`, `aedit_relation`, `acreate_entity`, `acreate_relation`, `amerge_entities`), before any lock is taken and before the first graph read, so a rejected payload has touched nothing. `ValueError` already maps to 400 on both routes, so no route change was needed.

Allowed fields:

| API | Fields |
|---|---|
| entity edit | `entity_name` (rename target), `entity_type`, `description`, `source_id`, `file_path` — strings |
| entity create / merge override | the same, minus `entity_name` |
| relation edit / create | `description`, `keywords`, `source_id`, `file_path` (strings), `weight` (number) |

The two families of caller differ deliberately in how they treat an unrecognised key:

- **edit / merge — reject (400).** These merge, so an unknown key *is written*. Silently dropping it on an edit endpoint would report success for a change that never happened. The error names the allowed fields.
- **create — keep ignoring it.** These already copy a fixed set of named fields, so an unknown key is structurally incapable of reaching storage. They only gain the type check, which avoids new 400s for payloads that work today.

`weight` accepts a numeric string (the create path has always run it through `float()`) and is normalized to `float`, so the *stored* attribute is a number on the edit path too, instead of a string that only the VDB payload converted.

**Contract layer** (`lightrag/utils.py`, `lightrag/base.py`) — `validate_graph_attributes` states the rule as the intersection of what the seven registered `GRAPH_STORAGE` backends can carry: identifier-shaped names, and values that are `str` (XML-compatible), `int` (int64), finite `float` or `bool`. `BaseGraphStorage.upsert_node` documents it as the caller's contract and points at the shared helper rather than inviting each backend to re-derive it.

The string rule is the XML 1.0 `Char` production, matched against GraphML's actual acceptance: tab, newline and carriage return stay legal, because descriptions carry all three.

> The *name* half of that rule is narrowed in the follow-up (#3624) from "identifier-shaped" to the actual interpretation hazard (no `.`, no leading `$`), after review pointed out that the wider rule would strand entities holding an unusual-but-harmless legacy attribute name. Nothing in this PR calls it — the entry layer works from the explicit field allowlist above — so the narrowing changes no behaviour here.

## What's deliberately not here

Enforcing the contract *inside* the backends: NetworkX validate-before-mutate (its "mutate, then serialize the whole graph, no rollback" shape is what turns one bad value into an instance-wide outage) and MongoDB attribute-name rejection (attribute names go straight into a `$set`, where a dot is a path separator and `source_ids.0` would rewrite an element of the chunk-attribution array). Those are backend-specific defects, not a change to the abstract contract, and they are a follow-up PR stacked on this one (#3624). The other four backends need no change — the driver or the DB already refuses, and the entry layer now covers the rest.

Also not here: length caps on attribute values. `MAX_REQUEST_BODY_BYTES` (1 MiB by default) bounds a single request; cumulative growth across many requests is a separate concern.

## Testing

```
./scripts/test.sh tests/api/routes tests/utils tests/kg/networkx_impl tests/extraction
767 passed, 4 skipped
```

New: 51 offline tests.

- `tests/api/routes/test_graph_attribute_validation.py` — behavioural fix proofs. Each drives a **real** `NetworkXStorage` and asserts an ordinary unrelated write still reaches disk after a hostile payload is refused. On the pre-fix tree these fail because that write raises `TypeError: GraphML does not support ...` — not because a symbol is missing; the module deliberately imports nothing added by this PR so it stays collectable there. Covers all four attack shapes above plus the merge path, and pins that a multi-line description is still accepted (guarding against over-rejection).
- `tests/utils/test_graph_attribute_validator.py` — the shared validator's own rules.

### Revised after review

Two P2 findings from Codex, both valid, both fixed:

- **Coerced `weight` was not revalidated** (f1839293). The value check ran on what the caller sent, and a numeric string is a perfectly storable `str` — it is the `float()` afterwards that manufactures the forbidden value, so `{"weight": "nan"}` was accepted. `json.dumps` emits a bare `NaN`, which PGTable's `jsonb` column rejects, so an intended 400 would have arrived as a 500.
- **Oversized integers returned 500** (94391b29). `{"weight": 10**400}` — `json.loads` yields an unbounded Python int, and `float()` raises `OverflowError`, which is neither `TypeError` nor `ValueError`. Beyond catching it, the gap behind it is fixed: integers are now bounded to **int64**, because that *is* the intersection the contract claims. Verified against the installed neo4j driver (`packstream/v1/__init__.py:117-121` raises `OverflowError` outside `[-2**63, 2**63)`), and GraphML declares int attributes `attr.type="long"` = `xsd:long` = int64. Only PostgreSQL's arbitrary-precision `jsonb numeric` was genuinely fine with a larger value.

These bounds are a **caller** contract only. A later review round showed that enforcing them inside a storage backend strands data — GraphML round-trips `NaN` and out-of-int64 integers, so a workspace can already hold one — so #3624 splits the value rule and has NetworkX enforce only what XML cannot encode. Nothing changes here: caller input still has to meet the full contract.

One existing test moved behaviour: `test_aedit_entity_rejects_empty_description` sends `{"description": None}`. The empty-description check now runs *before* sanitization so `None` keeps reporting itself as an empty description rather than as a type error — both refuse before storage is touched, so the ordering is about the message only. There is a comment at each call site to stop it being "tidied" back.

Note for reviewers reproducing the pre-fix failures: running the whole fix-proof file in one process on the pre-fix tree segfaulted for me while pytest was formatting the failures (macOS / CPython 3.12, lxml + spacy loaded). Individually each test fails cleanly, and a standalone script that fails the same GraphML write 400 times in one process does not crash — so this looks like a local harness artifact on a code path that no longer exists after the fix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/utils_graph.py [hunk 1]
```diff
@@ -11,13 +11,133 @@
...
+                raise ValueError(
+                    f"{object_type.capitalize()} field '{key}' must be a number, "
+                    "got bool"
+                )
+            try:
+                # A numeric string is accepted because the create paths have
+                # always run the value through `float()`; normalizing here means
+                # the *stored* attribute is a float on the edit paths too,
+                # instead of a string that only the VDB payload converted.
+                sanitized[key] = float(value)
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/utils_graph.py]: Confirmed and fixed in f1839293.

Reproduced first — `"nan"`, `"NaN"`, `"inf"`, `"Infinity"`, `"1e999"` and `"-1e999"` were all accepted and stored as non-finite floats, so the finite-number rule this PR documents was violated by the PR's own number path.

The cause is that the value check ran on what the caller sent, and for a numeric string that is a perfectly storable `str`; it is the `float()` conversion afterwards that manufactures the forbidden value. The coerced result now goes back through the same shared helper before it is stored.

Your consequence analysis is right and is what makes this more than a cosmetic inconsistency: `json.dumps` emits a bare `NaN`, which PGTableGraphStorage's `jsonb` column rejects, so an intended 400 would have arrived as a 500. Regression test: `test_relation_weight_rejects_strings_that_coerce_to_non_finite`.

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/utils_graph.py [hunk 2]
```diff
@@ -11,13 +11,146 @@
...
+                    f"{object_type.capitalize()} field '{key}' must be a number, "
+                    "got bool"
+                )
+            try:
+                # A numeric string is accepted because the create paths have
+                # always run the value through `float()`; normalizing here means
+                # the *stored* attribute is a float on the edit paths too,
+                # instead of a string that only the VDB payload converted.
+                coerced = float(value)
+            except (TypeError, ValueError):
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/utils_graph.py]: Confirmed and fixed in 94391b29. Reproduced first: `json.loads` yields an unbounded Python int, the value check accepted it as an `int`, and `float()` raised `OverflowError` — neither `TypeError` nor `ValueError`, so it escaped the handler and reached the route as a 500.

I took the `OverflowError` catch as well, but it only covers the one numeric field, so I fixed the gap behind it too. The value rule accepted integers of any size while claiming to be the intersection of what the registered backends can carry, and an int past int64 is not in that intersection:

- the neo4j driver's packstream `Packer` raises `OverflowError("Integer ... out of range")` outside `[-2**63, 2**63)` — read from the installed package (`neo4j/_codec/packstream/v1/__init__.py:117-121`), not assumed;
- GraphML declares an int attribute as `attr.type="long"`, and `xsd:long` is int64, so networkx writes a larger value under a type that cannot hold it;
- only PostgreSQL is genuinely fine with it, because `jsonb` numbers are arbitrary-precision `numeric`.

So integers are bounded to int64 by the shared rule, which refuses an oversized `weight` before the conversion is attempted. Tests: `test_relation_weight_rejects_oversized_integers`, `test_integers_outside_int64_are_rejected`, plus the int64 boundary values asserted accepted.

One boundary I did **not** soften, recorded in the commit message: on NetworkX a graph that already stores a non-finite float or an out-of-int64 integer — reachable only by having sent one through the pre-fix edit API, since GraphML round-trips both — will now refuse rewrites of that object, and needs it deleted and recreated. Unlike the attribute-name case in #3624, where the rejected names were legal on every backend, these values genuinely are not, so tolerating them would reopen the cross-backend divergence this rule exists to close.

**Comment by danielaskdd** [comment]
@codex review
